src/static/helpers/slasHelper.ts

```diff
@@ -0,0 +1,262 @@
+ * @param codeVerifier random string to use as a code verifier
+ * @returns code challenge
+ */
+export const generateCodeChallenge = async (
+  codeVerifier: string
+): Promise<string> => {
+  const urlSafe = (input: string) =>
+    input.replace(/\+/g, '-').replace(/\//g, '_').replace(/=/g, '');
+
+  let challenge = '';
+  try {
+    // Cannot test browser functions
+    /* istanbul ignore next */
+    if (window.crypto.subtle.digest !== undefined) {
+      const encoder = new TextEncoder();
+      const data = encoder.encode(codeVerifier);
+      const digest = await window.crypto.subtle.digest('SHA-256', data);
+      const base64Digest = btoa(String.fromCharCode(...new Uint8Array(digest)));
+      challenge = urlSafe(base64Digest);
+    }
+  } catch (e) {
+    const crypto = await import('crypto');
+    challenge = urlSafe(
+      crypto.default.createHash('sha256').update(codeVerifier).digest('base64')
+    );
+  }
+
+  /* istanbul ignore next */
+  if (challenge.length === 0) {
+    throw new Error('Problem generating code challenge');
+  }
+
+  return challenge;
+};
+
+/**
+ * Wrapper for the authorization endpoint. For federated login (3rd party IDP non-guest), the caller should redirect the user to the url in the url field of the returned object. The url will be the login page for the 3rd party IDP and the user will be sent to the redirectURI on success. Guest sessions return the code and usid directly with no need to redirect.
+ * @param slasClient a configured instance of the ShopperLogin SDK client
+ * @param parameters -
+ *   codeVerifier - random string created by client app to use as a secret in the request
+ *   redirectURI - the location the client will be returned to after successful login with 3rd party IDP. Must be registered in SLAS.
+ *   hint? - optional string to hint at a partitcular IDP. Guest sessions are created by setting this to 'guest'
+ *   usid? - optional saved SLAS user id to link the new session to a previous session
+ * @returns login url, user id and authorization code if available
+ */
+export async function authorize(
+  slasClient: ShopperLogin<{
+    shortCode: string;
+    organizationId: string;
+    clientId: string;
+  }>,
+  parameters: {
+    codeVerifier: string;
+    redirectURI: string;
+    hint?: string;
+    usid?: string;
+  }
+): Promise<Record<string, string>> {
+  const codeChallenge = await generateCodeChallenge(parameters.codeVerifier);
+
+  const options = {
+    parameters: {
+      client_id: slasClient.clientConfig.parameters.clientId,
+      code_challenge: codeChallenge,
+      ...(parameters.hint && {hint: parameters.hint}),
+      organizationId: slasClient.clientConfig.parameters.organizationId,
+      redirect_uri: parameters.redirectURI,
+      response_type: 'code',
+      ...(parameters.usid && {usid: parameters.usid}),
+    },
+  };
+
+  const response = await slasClient.authorizeCustomer(options, true);
```

Comment:
To improve server-side peformance here, you want to split client and server side handling.

The SLAS API under `authorizeCustomer` returns an HTTP 303 redirect to the callback URL in the `Location` header.

Server-side, you'd rather not follow the redirect (skip making another HTTP request) as you can dig the USID and code values you need out of `Location` header.

Client-side, you must follow the redirect because the fetch spec says that redirects are opaque and you cannot access the header: https://fetch.spec.whatwg.org/#atomic-http-redirect-handling

I've tried to do this poorly in the PWA Kit:

https://github.com/SalesforceCommerceCloud/pwa-kit/pull/254/files
https://github.com/SalesforceCommerceCloud/pwa-kit/blob/f300639e26453e17c755f208c93ec5d68f788390/packages/pwa/app/commerce-api/auth.js#L272-L290

Reply:
I'll need to play with the server-side a little which I'll try to do today. On the client side the behavior is:
1. Fetch call is made with default 'follow'
2. Server returns 303
3. A call is made to the redirect url by the browser (visible in network tab)
4. We ignore that and instead access the `url` property of the initial 303 response from the server and grab the code and usid
You can add the 'manual' option to avoid the browser making the redirect call, but then you aren't able to read anything from the response including that `url` property. I can see from the link you put that we should be able to use the 'manual' option on the server and still get the url so I'll try that, thanks!

Reply:
Done! conditionally setting 'manual' for serverside calls